### DISCUSSION TITLE
Ray tracing support

### DISCUSSION THREAD

**Discussion by rntrtul** [godotengine/godot-proposals#5162] [3D]
There's been previous discussion of adding ray tracing support to Godot ([in 2019](https://github.com/godotengine/godot/issues/27355) , [and more recently](https://github.com/godotengine/godot-proposals/issues/4870)). 

Vulkan seem to have added full support in their SDK along with validation layers and other helpful info [link to press release](https://www.khronos.org/news/press/vulkan-sdk-tools-and-drivers-are-ray-tracing-ready). So I think development or planning implementation could begin now. 

Another issue that was brought up was lack of hardware supporting ray tracing, but looking at the steam hardware survey from  [july 2022](https://store.steampowered.com/hwsurvey) I think 30% of GPU's support some form of hardware RT. 
* Ampere (RTX 30XX):  17.41%
* Turing (RTX 20XX): 11.66%
* RDNA2 (RX 6XXX): 1.32%

Now not all of these are using ray tracing or should be running ray tracing, but it's still a significant number. There is also the consoles that support ray tracing now. Accounting for the time it will take to implement ray tracing and then the time it will take for games to ship with RT, I think this total number will be higher.

some features for ray tracing are
* RT shadows
* RT reflections
* RT global illumination

Hopefully this can be used to discuss the RT implementation and plan going forward.

**Comment by Calinou** [comment]
> So I think development or planning implementation could begin now.

The current goal to stabilize the existing Vulkan renderer for 4.0. Someone was recently interested in exploring raytracing support in Godot as a side project, but I don't think it should be the focus *right now* unless you're doing this as a learning exercise. Feature freeze has also started for 4.0, so any new feature development will be targeted for 4.1 unless it's critical for 4.0.

Don't worry, the current Vulkan renderer will have plenty of features to keep a RTX 3090 Ti (and soon a RTX 4090) busy :slightly_smiling_face:

**Comment by Mercesa** [comment]
I am the person Calinou mentioned that's interested in adding RT support :). I have started evaluating Godot's structures on the best ways could be to support Raytracing within Godot in my spare time.

I'm currently in the progress of adding RT support for the `RenderingDevice` and calling the `RenderingDevice` from GdScript. 

I am planning to make a proposal at some point, but first want a working prototype to discover all the different types of problems I could run into before starting the proposal. I think the best focus would be first on adding support for rendering engineers to be able to use raytracing where they please through inline tracing, you could easily do shadows with some basic RT features.

There are two ways to trace rays:
- RT pipelines: A bespoke pipeline fully focused on raytracing, this would be dispatched by using `vkCmdTraceRaysKHR` similar to a compute shader is using dispatches.
- Inline raytracing: It is possible to trace rays from any type of shader stage (Compute, Frag, Vertex, etc). This is much easier to implement and support.

Both RT pipelines and Inline RT require acceleration structure building, but inline RT is definitely less effort to implement.

Adding RT is non-trivial, there are several layers of Godot that would require slight/big reworks/additions, I will describe a few elements of the cuff in no order of importance and will probably edit this post or add more posts as I gain more info. I am new to Godot's source, so I will definitely not have the full picture yet :).

- **Acceleration structure building (AS)**: This requires geometry information about the whole scene, and some type of geometry LOD'ing to make the acceleration structure building workable, since building all geometry as is, is not possible for runtime at runtime framerates with complex game scenes, so we would need some type of geometry simplification step. To make this engine-worthy, we would need to support BLAS/TLAS refitting and have dynamic updates within the acceleration structures so animations are also supported.
- **Shader authoring/system**: Godot has several layers of shading systems, users can make their own nodes, and raytracing shaders are quite complicated and generally are grouped together. At the moment I don't see an easy way to have users create their own RT shaders without having to either make their own renderpasses and integrate them with the current rendering systems.
- **Material complexity**: Engines that deal with lots of materials such as Godot, Unreal etc. can be quite difficult for Raytracing in terms of grouping. Every Material would need to have its own bespoke RT Hit shader, and this would in the end turn into a giant pipeline that can slow down PSO creation to a crawl.

**Reply by Ansraer** [reply]
iirc reduz' SDFGI already builds a simplified representation of the static geometry. You might want to look into whether its possible to extend that code to also build the BVH for static meshes at the same time.

**Reply by Calinou** [reply]
> I'm currently in the progress of adding RT support for the RenderingDevice and calling the RenderingDevice from GdScript.

@Mercesa Did you find time to progress on this? If you didn't manage to complete it, that's OK, but please make sure to post the unfinished code somewhere so that someone else may be able to finish it.

**Comment by mrjustaguy** [comment]
2 RT Related References to add from Godot Contributors Chat:
https://chat.godotengine.org/channel/rendering?msg=rn2NZeMbxNJs5gW4K
https://chat.godotengine.org/channel/rendering?msg=gnYbLKXZFKc7NEwGB

**Comment by mrjustaguy** [comment]
Another relevant link for this discussion
MIT Unity Realtime Path tracer (Has Compute Fallback for no Hardware RT)
https://github.com/Pjbomb2/TrueTrace-Unity-Pathtracer

**Reply by Calinou** [reply]
Note for future readers: a pathtracer is designed to work with offline rendering/Movie Maker mode, as opposed to being used during gameplay. The tradeoff is that it offers more accurate rendering compared to hybrid RT, as (almost) everything you see on screen is the result of raytracing.

Ideally, we should support both a hybrid RT approach and pathtracing. While hybrid RT is most suited for gameplay, pathtracing is an useful development tool as it can be used to confirm your scene's lighting is physically accurate. It can also be used to create accurate renderings with Movie Maker mode, which is useful for things like archviz.

When coupled with a denoiser, path tracing is also starting to become viable for gaming (though only on high-end GPUs currently). Future generations of GPUs are expected to make this more viable, especially in laptops and mid-range GPUs.

**Comment by radiantgurl** [comment]
Bumping thread as there is now a new cinematic renderer being proposed and there are still people wanting full raytracing support inside Godot.

**Reply by HydrogenC** [reply]
Is the ciematic renderer planned to adopt a Path Tracing approach which is not viable for realtime, or a hybrid Rasterization-RT approarch? Where could I find more technical information about the proposed cinematic renderer (along with the proposed deferred renderer)? Thanks in advance.

**Reply by octanejohn** [reply]
the correct way is a new renderer with amd gi 1.1/path tracing but a hybrid way with the current renderer for shadows/ao/reflection maybe wanted

https://gist.github.com/reduz/c5769d0e705d8ab7ac187d63be0099b5

https://github.com/godotengine/godot-proposals/issues/8295

**Reply by HydrogenC** [reply]
> the correct way is a new renderer with amd gi 1.1/path tracing but a hybrid way with the current renderer for shadows/ao/reflection maybe wanted
> 
> https://gist.github.com/reduz/c5769d0e705d8ab7ac187d63be0099b5
> 
> #8295

Thanks for your information. A fully raytraced pipeline is more cutting edge, but a hybrid way seems more mainstream, at least in present. May be out of performance concerns.

**Comment by radiantgurl** [comment]
STATUS UPDATE: Work started on implementing hardware raytracing support. https://github.com/godotengine/godot/pull/99119

**Reply by RC-1290** [reply]
And it looks like it has been merged into Godot Main!

It's fairly low level, so it's not a button to turn on ray tracing, and everything looks fancy. But it does let you create a ray tracing set-up only using Gdscript and glsl shaders. There are some limitations, but you can definitely make some interesting projects with this, if you're okay with manually feeding the objects you want to ray trace to the ray tracing pipeline.

I believe one of the limitations is that you can't use a shader binding table, so you kind of have to fit everything into a single shader.

**Reply by Calinou** [reply]
Godot currently doesn't support bindless rendering, so in practice, I expect RTAO and RT shadows to be feasible, but RT reflections and RT global illumination need bindless rendering to really be viable. With RT reflections and RT GI, you need to read the material of each object to implement the respective effect, while RT shadows and RTAO can work with an universal material (for opaque objects at least – alpha cut-out will require a different texture, so it's probably not possible either).

**Comment by zun1uwu** [comment]
I'm a beginner, but I recently came across SVOGI (Sparse Voxel Octree Global Illumination) through [Crytek's Neon Noir Benchmark (YouTube)](https://www.youtube.com/watch?v=1nqhkDm2_Tw) and ran the benchmark myself afterwards.

I run a Ryzen 5 3600 and a GTX 1070 Ti and was surprised by the results, because I had an average FPS of around 80 when running the Benchmark on Utra.

So this method of ray tracing has seemingly amazing performance and I think this could potentially be the route Godot could go when implementing something like ray tracing or a great addition to "normal" ray tracing, especially since it runs on machines without specialized RT cores.

But since everyone is talking about ray-tracing with DLSS and other AI upscaling methods to improve the performance of traditional polygon-based ray tracing and SVOGI seems to be kind of "forgotten" (note that the tech demo is 6 years old), there has to be a catch, right?

**Reply by mrjustaguy** [reply]
there is a catch. as a Voxel based approach it start's falling apart really fast in quality with things that aren't close to the camera, and cannot represent some types of complex environments well, the way RT can.

The only problem of RT is that it requires at least a few decades of Hardware and Software improvements to not be an ultra noisy, boiling, upscaling/fake frame required mess.

One thing I've learnt from years in game dev, there's always a compromise to make, and it's all about picking the right ones to make.

**Reply by Calinou** [reply]
The idea of SVOGI is pretty close to Godot's VoxelGI and SDFGI, which are methods to achieve global illumination without raytracing by using coarser approximation structures (voxels and signed distance fields respectively).

Hardware-accelerated raytracing can get much more precise because it works with triangles, not voxels or SDFs. Even though you need to build an acceleration structure (BVH) with simplified geometry and limited draw distance, it will basically always be more precise than voxels or SDFs.

**Reply by zun1uwu** [reply]
The voxel ray tracing simulation in the benchmark provided reflections that looked extremely detailed (just like the video shows) and very much comparable to traditional polygon ray tracing **(which baffled me)** in comparison to the ones generated by SDFGI or Godot's VoxelGI, hence I was wondering.

**Reply by mrjustaguy** [reply]
IIRC that's because at like a few meters of distance they're using short ray tracing for smooth surface reflections, on lower LOD versions of it.

Other engines typically just switch to SSR and use Voxel reflections as a fallback (like Lumen)

**Reply by zun1uwu** [reply]
Very interesting, thank you for explaining it. Does it also explain why it runs so well on my GTX 1070 Ti? I hope I'm not being bothersome, it's my first time writing in a thread like this (I still feel a little bit clueless about this).

**Reply by mrjustaguy** [reply]
Yeah, Voxel based GI and SSR are typically quite light, and Low levels of actual Ray Tracing can also be fairly light, it's just that such Low levels of Ray Tracing are typically unusable for most things outside of like basic non soft shadows and basic smooth surface reflections

**Comment by TheAenema** [comment]
Hi, I’ve developed a real-time path tracer for Godot 4 using latest NVIDIA path tracing technology, DLSS 4.0 (RR) and it doesn't require any changes to the engine. It has the same performance as publicly released NVIDIA path tracer and has zero blit and overhead.
Here's [a video](https://www.youtube.com/watch?v=TkPs0EsCt7E) showcasing it in action.

![JenovaRTXShot1](https://github.com/user-attachments/assets/2cd61cd5-d558-48a4-80d2-b3a940efff4c)
![JenovaRTXShot2](https://github.com/user-attachments/assets/6cf4ef81-48ce-411f-b0b8-7a3e96b16b51)
![JenovaRTXShot3](https://github.com/user-attachments/assets/569f02d1-ac55-4b54-80e9-136d1ac27b97)
![JenovaRTXShot4](https://github.com/user-attachments/assets/991d092e-5794-406e-8845-7d460b9d8bbd)
![image-13](https://github.com/user-attachments/assets/6f6a3e55-afa4-4765-abeb-269a5b5dec9c)
![image-14](https://github.com/user-attachments/assets/87e232af-c818-4f51-a6c9-431da931b961)

My implementation is entirely developed in [Jenova Framework](https://github.com/Jenova-Framework) and it works on Godot 4.3+ & Windows 10/11 x64, It's very flexible and can be used to render whole scene or just lights, shadows and reflections.

I'm still adding features but I'm pretty happy with the results.

**Reply by Sasino97** [reply]
It looks really amazing. Do you plan on releasing it? Or at the very least selling licenses to use it? Or is it just your personal project?

**Comment by arlack9** [comment]
unlike the other engines were new versions are sluggish to run on a 4gb vram , bcoz of feature integration into engines , countless shader compilations ,
godot needs to be simple portable , including the c# workflow, 

what i suggest is c# nuget packages contain some low level bindings for vulkan that can somehow reduce complication in feature development 

Silk.NET: Vulkan/OpenGL bindings for custom pipelines, usable for ray marching dispatchers.

ComputeSharp: HLSL compute shaders on DirectX12/Vulkan, adaptable for path tracing kernels.

Raylib-cs: Simplified 3D rendering with basic raycasting, though not full RT.

loosely coupled integration is better when dealing with new integrations as it reduces code complexity.

**Reply by mrjustaguy** [reply]
that's... not the root cause of why they run poorly.

**Comment by Ansraer** [comment]
For everyone's information, https://github.com/godotengine/godot/pull/99119 was just merged.

Note that this does **NOT** mean that Godot now actually uses raytracing. However, with this PR Godot now has the necessary API to make it possible for developers to add raytracing features either by forking the engine or using "simple" gdscripts.

Many thanks to the author @Fahien and all the rendering engineers who took the time to thoroughly review and give feedback for the PR in question. :tada:

**Comment by FeralBytes** [comment]
I think this work by NVIDIA is relevant to this conversation: https://github.com/NVIDIA-RTX/godot

**Reply by RC-1290** [reply]
Do they have any public information about this, besides the repo? Because I'm wondering whether they're making use of the new Vulkan ray tracing support of 4.7. Perhaps it's mostly meant as a proof of concept? I heard there was a GDC presentation about it, but that usually takes a while to appear in the vault.

**Reply by Saul2022** [reply]
They are using the vulkan stuff from the API

**Reply by FeralBytes** [reply]
@RC-1290 I saw a single slide of a slide presentation, but beyond that I was not able to find anything else relevant out about it.

**Reply by mrjustaguy** [reply]
It's using the new Vulkan RT Plumbing yes.

It's still WIP (AMD for example doesn't work, the Nvidia Engineer will investigate that next week) and there's a bunch of things that it runs into issues with due to not being implemented or improperly implemented, such as Vertex Shaders and Skinned Meshes.

It was rushed out due to GDC deadline, and AI (Cursor) was used.

**Reply by ynot01** [reply]
the README is unmodified, does anyone know how to use it? does the GDC slide explain it?

**Reply by mrjustaguy** [reply]
build the dlss dev branch, it's in the environment settings under pathtracing

**Reply by ynot01** [reply]
as a heads up to others, it also doesn't support linux yet, but i imagine that will change eventually (if they intend to keep going) given godot's multi-platform philosophy

**Reply by Calinou** [reply]
This proof-of-concept branch supports Linux: https://github.com/plankatron/godot/tree/ngx-dlss-linux

It doesn't use Streamline but uses direct NGX integration for DLSS SR/RR support (FG is not supported in native Linux apps as far as I know). The final implementation will likely be significantly different, but I don't know the specifics.

Fun fact: While Baldur's Gate 3 beat Godot to native Linux support for DLSS SR, this is the very first publicly available native Linux app to support DLSS RR 🙂

**Comment by pockerhead** [comment]
Just shipped an experimental RT lighting stack (RT shadows with PCSS penumbra, RTAO,
1-bounce GI + SVGF-style denoiser) on the 4.7 RenderingDevice API in a procedural
open-world game — huge thanks for #99119, it holds up in production use. Real-scene
numbers: the whole RT stack costs ~0.8 ms where cascaded shadow maps cost ~1.85 ms.

Walls we hit, in order of pain:
1. BLAS refit/update not exposed → no animated geometry in RT (filed a proposal: [godot-proposals#15180](https://github.com/godotengine/godot-proposals/issues/15180));
2. no compaction (same proposal);
3. `SUPPORTS_RAY_QUERY` can never be true — the feature is not requested at device creation (filed: [godot#121286](https://github.com/godotengine/godot/issues/121286));
4. push constants arrive as zeros in raygen on NVIDIA — root cause identified in the issue (stage enum ordinals ORed instead of `_BIT` flags in the reflection, so `VkPushConstantRange` carries no RT stages; looks like a one-line fix), workaround is a UBO (filed: [godot#121285](https://github.com/godotengine/godot/issues/121285));
5. longer-term: `VK_KHR_ray_tracing_position_fetch` (we duplicate vertex data into SSBOs to shade hits) and opacity micromaps — `VK_EXT_opacity_micromap`, now promoted to `VK_KHR_opacity_micromap` in Vulkan 1.4.351 (any-hit alpha-tested vegetation is the classic worst case; Indiana Jones reports −55% vegetation trace cost with OMM).

raster
<img width="1280" height="719" alt="photo_2026-07-12_03-13-08 (2)" src="https://github.com/user-attachments/assets/2ef40805-bbd2-4155-a299-818a871f7876" />
RT
<img width="1280" height="720" alt="photo_2026-07-12_03-13-08" src="https://github.com/user-attachments/assets/a3606691-15aa-40ba-9fbd-43cd4ad8288c" />
raster
<img width="1280" height="720" alt="photo_2026-07-12_17-23-51" src="https://github.com/user-attachments/assets/ccabe1c8-516b-49d1-874e-e683fc57ac36" />
RT
<img width="1280" height="722" alt="photo_2026-07-12_19-05-16" src="https://github.com/user-attachments/assets/a81e038d-9b9c-4de8-9d63-70ac98e328c8" />

**Comment by coffeel0RD** [comment]
I’m new to godot but was wanting to have raytraced reflections how hard would it be to implement? I dont want full pathtracing and don't want it to only run on nivida cards

**Reply by Ansraer** [reply]
Tricky I would say. The problem ist that you need a representation of your scene on the gpu in a format that ray tracing can understand, which the engine currently doesn't provide for scenes. So you would need to build this data yourself, and once you have the scene data on the gpu you are already more than halfway there to a full pathtracer anyways.

**Reply by mrjustaguy** [reply]
The Path tracer runs on non Nvidia too, it just lacks denoising atm on non Nvidia cards

**Reply by coffeel0RD** [reply]
The issue is pathtracing has such a performance hit

**Reply by coffeel0RD** [reply]
> The Path tracer runs on non Nvidia too, it just lacks denoising atm on non Nvidia cards

Interesting

**Comment by ttimpe** [comment]
Is anyone working on adding proper raytracing support for Apple M-series chips using Metal yet?

**Reply by ejektaflex** [reply]
My understanding is that this relies on Molten VK adding the Vulkan ray tracing extensions, which has not happened yet.

It sounds like MoltenVK has been in maintenance mode for a while, so it seems like that feature has no definite timeline, but I'd be happy for someone else to correct me on that.

**Reply by ttimpe** [reply]
I'm thinking about just throwing Claude at it, but I honestly don't want to spend that amount of money. I'd probably be better off just buying a PC with a good GPU. But would there even be a price difference at this point?

**Reply by Calinou** [reply]
> My understanding is that this relies on Molten VK adding the Vulkan ray tracing extensions, which has not happened yet.

Godot has native Metal support (which is recommended to use on all Apple Silicon hardware), so we don't need to wait for MoltenVK to add support for it. Adding support for Metal raytracing to the Metal driver will likely be less work, and also more efficient.

cc @stuartcarnie
